# Motor Vehicle Collisions -- Data Profiling and Data Cleaning

## Dataset and Streams

The [Motor Vehicle Collisions](https://data.cityofnewyork.us/Public-Safety/Motor-Vehicle-Collisions-Crashes/h9gi-nx95) dataset contains information from all police reported motor vehicle collisions in NYC. Each row represents a crash event. 

The dataset consists of 1.84 million rows and uncompressed data file is about 377MB in size. 

**[OpenClean](https://github.com/VIDA-NYU/openclean)** is a Python library for data profiling and data cleaning. It is primarily intended for tabular datasets, which are represented as pandas data frames. After downloading the dataset file, we can, for example, open it as a data frame using the pandas.read_csv() function. This requires reading the full dataset into main memory. For large datasets this may not be feasible, and in some cases we don’t need all the data. For these scenarios, openclean’s data streaming operator allows us to apply data profiling, filtering, and transformation operations directly on a stream of rows that are read from the file. 

In [1]:
# Due to the size of the full dataset file, we make use of openclean's
# stream operator to avoid having to load the dataset into main-memory.

from openclean.pipeline import stream

file_path = 'Motor_Vehicle_Collisions_Crashes.csv'
ds = stream(file_path)

In [2]:
# Count number of records in the dataset.

print(f'{ds.count():,} rows.')

1,839,256 rows.


In [3]:
# Print the first ten rows of the dataset to get a first idea of the content.

ds.head()

CRASH DATE CRASH TIME   BOROUGH ZIP CODE  LATITUDE  LONGITUDE  \
0  04/14/2021       5:32                                           
1  04/13/2021      21:35  BROOKLYN    11217  40.68358  -73.97617   
2  04/15/2021      16:15                                           
3  04/13/2021      16:00  BROOKLYN    11222                        
4  04/12/2021       8:25                            0          0   
5  04/13/2021      17:11                                           
6  04/13/2021      17:30    QUEENS    11106                        
7  04/16/2021      23:30                                           
8  04/11/2021      17:00                                           
9  04/16/2021      21:15                                           

                LOCATION                    ON STREET NAME CROSS STREET NAME  \
0                                  BRONX WHITESTONE BRIDGE                     
1  (40.68358, -73.97617)                                                       
2                                 HUTCHINSON RIVER PARKWAY                     
3                                        VANDERVORT AVENUE    ANTHONY STREET   
4             (0.0, 0.0)  EDSON AVENUE                                         
5                                   VERRAZANO BRIDGE UPPER                     
6                                                    33 st             31ave   
7                                            SHORE PARKWAY                     
8                         GOWANUS RAMP                                         
9                                 BRONX RIVER PARKWAY RAMP                     

                            OFF STREET NAME  ...  \
0                                            ...   
1  620       ATLANTIC AVENUE                 ...   
2                                            ...   
3                                            ...   
4                                            ...   
5                                            ...   
6                                            ...   
7                                            ...   
8                                            ...   
9                                            ...   

  CONTRIBUTING FACTOR VEHICLE 2 CONTRIBUTING FACTOR VEHICLE 3  \
0                   Unspecified                                 
1                                                               
2                                                               
3                   Unspecified                                 
4                   Unspecified                                 
5                   Unspecified                                 
6                   Unspecified                                 
7                                                               
8               Other Vehicular                                 
9                   Unspecified                                 

  CONTRIBUTING FACTOR VEHICLE 4 CONTRIBUTING FACTOR VEHICLE 5 COLLISION_ID  \
0                                                                  4407480   
1                                                                  4407147   
2                                                                  4407665   
3                                                                  4407811   
4                                                                  4406885   
5                                                                  4407883   
6                                                                  4408019   
7                                                                  4408060   
8                                                                  4406314   
9                                                                  4408149   

                   VEHICLE TYPE CODE 1 VEHICLE TYPE CODE 2  \
0                                Sedan               Sedan   
1                                Sedan                       
2  Station Wagon/Sport Utility Vehicle 

# Data Profiling

Data profiling refers to the activity of creating small but informative summaries of a database. Profiling helps users to gain an understanding of the data properties and to uncover data quality flaws.

Openclean supports a variety of different data profiling operators.

In [4]:
# Profile the dataset view using the default data profiler.

from openclean.profiling.column import DefaultColumnProfiler

profiles = ds.profile(default_profiler=DefaultColumnProfiler)

In [5]:
# Print overview of profiling results.

profiles.stats()

total    empty  distinct  uniqueness  \
CRASH DATE                     1839256        0      3418    0.001858   
CRASH TIME                     1839256        0      1440    0.000783   
BOROUGH                        1839256   568098         5    0.000004   
ZIP CODE                       1839256   568320       232    0.000183   
LATITUDE                       1839256   214986    122315    0.075305   
LONGITUDE                      1839256   214986     95951    0.059073   
LOCATION                       1839256   214986    243078    0.149654   
ON STREET NAME                 1839256   375915     15967    0.010911   
CROSS STREET NAME              1839256   661099     19238    0.016329   
OFF STREET NAME                1839256  1555710    178692    0.630205   
NUMBER OF PERSONS INJURED      1839256       18        27    0.000015   
NUMBER OF PERSONS KILLED       1839256       31         7    0.000004   
NUMBER OF PEDESTRIANS INJURED  1839256        0        13    0.000007   
NUMBER OF PEDESTRIANS KILLED   1839256        0         4    0.000002   
NUMBER OF CYCLIST INJURED      1839256        0         5    0.000003   
NUMBER OF CYCLIST KILLED       1839256        0         3    0.000002   
NUMBER OF MOTORIST INJURED     1839256        0        27    0.000015   
NUMBER OF MOTORIST KILLED      1839256        0         6    0.000003   
CONTRIBUTING FACTOR VEHICLE 1  1839256     5306        61    0.000033   
CONTRIBUTING FACTOR VEHICLE 2  1839256   266317        61    0.000039   
CONTRIBUTING FACTOR VEHICLE 3  1839256  1713307        50    0.000397   
CONTRIBUTING FACTOR VEHICLE 4  1839256  1811717        39    0.001416   
CONTRIBUTING FACTOR VEHICLE 5  1839256  1831954        29    0.003972   
COLLISION_ID                   1839256        0   1839256    1.000000   
VEHICLE TYPE CODE 1            1839256    10092      1310    0.000716   
VEHICLE TYPE CODE 2            1839256   316452      1438    0.000944   
VEHICLE TYPE CODE 3            1839256  1717051       202    0.001653   
VEHICLE TYPE CODE 4            1839256  1812543        84    0.003145   
VEHICLE TYPE CODE 5            1839256  1832155        55    0.007745   

                                 entropy  
CRASH DATE                     11.676185  
CRASH TIME                      8.928377  
BOROUGH                         2.118321  
ZIP CODE                        7.221118  
LATITUDE                       15.631652  
LONGITUDE                      15.342183  
LOCATION                       16.179236  
ON STREET NAME                 10.581327  
CROSS STREET NAME              11.808005  
OFF STREET NAME                16.903913  
NUMBER OF PERSONS INJURED       0.965950  
NUMBER OF PERSONS KILLED        0.014318  
NUMBER OF PEDESTRIANS INJURED   0.298479  
NUMBER OF PEDESTRIANS KILLED    0.008107  
NUMBER OF CYCLIST INJURED       0.163679  
NUMBER OF CYCLIST KILLED        0.001464  
NUMBER OF MOTORIST INJURED      0.756799  
NUMBER OF MOTORIST KILLED       0.006323  
CONTRIBUTING FACTOR VEHICLE 1   3.519120  
CONTRIBUTING FACTOR VEHICLE 2   1.229855  
CONTRIBUTING FACTOR VEHICLE 3   0.592786  
CONTRIBUTING FACTOR VEHICLE 4   0.495097  
CONTRIBUTING FACTOR VEHICLE 5   0.495551  
COLLISION_ID                   20.810691  
VEHICLE TYPE CODE 1             3.264982  
VEHICLE TYPE CODE 2             3.570997  
VEHICLE TYPE CODE 3             2.915826  
VEHICLE TYPE CODE 4             2.732442  
VEHICLE TYPE CODE 5             2.684250

Combining profiling results with columns introduction in the [website](https://data.cityofnewyork.us/Public-Safety/Motor-Vehicle-Collisions-Crashes/h9gi-nx95), we know that the Motor Vehicle Collisions dataset has 29 columns.

* CRASH DATE: Occurrence date of collision. It does not have empty values.
* CRASH TIME: Occurrence time of collision. It does not have empty values.
* BOROUGH: Borough where collision occurred. It has 568098 empty values. It has 5 distinct values.
* ZIP CODE: Postal code of incident occurrence. It has 568320 empty values.
* LATITUDE: Latitude coordinate for Global Coordinate System, WGS 1984, decimal degrees (EPSG 4326). It has 214986 empty values.
* LONGITUDE: Longitude coordinate for Global Coordinate System, WGS 1984, decimal degrees (EPSG 4326). It has 214986 empty values.
* LOCATION: Latitude, Longitude pair. It has 214986 empty values.
* ON STREET NAME: Street on which the collision occurred. It has 375915 empty values.
* CROSS STREET NAME: Nearest cross street to the collision. It has 661099 empty values.
* OFF STREET NAME: Street address if known. It has 1555710 empty values.		
* NUMBER OF PERSONS INJURED: Number of persons injured.
* NUMBER OF PERSONS KILLED: Number of persons killed.	
* NUMBER OF PEDESTRIANS INJURED: Number of pedestrians injured.		
* NUMBER OF PEDESTRIANS KILLED:	Number of pedestrians killed.
* NUMBER OF CYCLIST INJURED: Number of cyclists injured.	
* NUMBER OF CYCLIST KILLED:	Number of cyclists killed.
* NUMBER OF MOTORIST INJURED: Number of vehicle occupants injured.
* NUMBER OF MOTORIST KILLED: Number of vehicle occupants killed.
* CONTRIBUTING FACTOR VEHICLE 1,2,3,4,5: Factors contributing to the collision for designated vehicle.	
* COLLISION_ID: Unique record code generated by system. Primary Key for Crash table. It does not have empty values.
* VEHICLE TYPE CODE 1,2,3,4,5: Type of vehicle based on the selected vehicle category.

In [6]:
profiles.types()

date   float      int     str
CRASH DATE                     3418       0        0       0
CRASH TIME                        0       0        0    1440
BOROUGH                           0       0        0       5
ZIP CODE                          0       0      231       1
LATITUDE                          0  122314        1       0
LONGITUDE                         0   95949        2       0
LOCATION                          0       0        0  243078
ON STREET NAME                   86       0       16   15865
CROSS STREET NAME                 2       1       26   19209
OFF STREET NAME                  30       0        1  178661
NUMBER OF PERSONS INJURED         0       0       27       0
NUMBER OF PERSONS KILLED          0       0        7       0
NUMBER OF PEDESTRIANS INJURED     0       0       13       0
NUMBER OF PEDESTRIANS KILLED      0       0        4       0
NUMBER OF CYCLIST INJURED         0       0        5       0
NUMBER OF CYCLIST KILLED          0       0        3       0
NUMBER OF MOTORIST INJURED        0       0       27       0
NUMBER OF MOTORIST KILLED         0       0        6       0
CONTRIBUTING FACTOR VEHICLE 1     0       0        2      59
CONTRIBUTING FACTOR VEHICLE 2     0       0        2      59
CONTRIBUTING FACTOR VEHICLE 3     0       0        2      48
CONTRIBUTING FACTOR VEHICLE 4     0       0        0      39
CONTRIBUTING FACTOR VEHICLE 5     0       0        0      29
COLLISION_ID                      0       0  1839256       0
VEHICLE TYPE CODE 1               0       0       12    1298
VEHICLE TYPE CODE 2               0       0       12    1426
VEHICLE TYPE CODE 3               0       0        0     202
VEHICLE TYPE CODE 4               0       0        0      84
VEHICLE TYPE CODE 5               0       0        0      55

This dataset has some type errors.

In [7]:
# Print the most frequent data type for each column.

print('Schema\n------')
for col in ds.columns:
    p = profiles.column(col)
    print("  '{}' ({})".format(col, p['datatypes']['distinct'].most_common(1)[0][0]))

Schema
------
  'CRASH DATE' (date)
  'CRASH TIME' (str)
  'BOROUGH' (str)
  'ZIP CODE' (int)
  'LATITUDE' (float)
  'LONGITUDE' (float)
  'LOCATION' (str)
  'ON STREET NAME' (str)
  'CROSS STREET NAME' (str)
  'OFF STREET NAME' (str)
  'NUMBER OF PERSONS INJURED' (int)
  'NUMBER OF PERSONS KILLED' (int)
  'NUMBER OF PEDESTRIANS INJURED' (int)
  'NUMBER OF PEDESTRIANS KILLED' (int)
  'NUMBER OF CYCLIST INJURED' (int)
  'NUMBER OF CYCLIST KILLED' (int)
  'NUMBER OF MOTORIST INJURED' (int)
  'NUMBER OF MOTORIST KILLED' (int)
  'CONTRIBUTING FACTOR VEHICLE 1' (str)
  'CONTRIBUTING FACTOR VEHICLE 2' (str)
  'CONTRIBUTING FACTOR VEHICLE 3' (str)
  'CONTRIBUTING FACTOR VEHICLE 4' (str)
  'CONTRIBUTING FACTOR VEHICLE 5' (str)
  'COLLISION_ID' (int)
  'VEHICLE TYPE CODE 1' (str)
  'VEHICLE TYPE CODE 2' (str)
  'VEHICLE TYPE CODE 3' (str)
  'VEHICLE TYPE CODE 4' (str)
  'VEHICLE TYPE CODE 5' (str)


### Explore some data properties

In [8]:
# Print the minimum and maximum value for column 'Issue Date'

profiles.minmax('CRASH DATE')

min        max
date 2012-07-01 2021-11-08

We could know that this dataset records crash events happening from 07/01/2012 to 11/08/2021.

In [9]:
# Print the most frequent values in column 'Violation Time'

profiles.column('CRASH TIME').get('topValues')

[('16:00', 26243),
 ('17:00', 25720),
 ('15:00', 25468),
 ('18:00', 23753),
 ('14:00', 23378),
 ('13:00', 21743),
 ('9:00', 19602),
 ('19:00', 19564),
 ('12:00', 19504),
 ('16:30', 19035)]

It seems that motor vehicle crash events are likely to happen during Rush Hour.

In [10]:
# Print the most frequent values in column 'BOROUGH'

profiles.column('BOROUGH').get('topValues')

[('BROOKLYN', 400172),
 ('QUEENS', 340459),
 ('MANHATTAN', 292471),
 ('BRONX', 184511),
 ('STATEN ISLAND', 53545)]

It seems that BOROUGH has no typos.

## Data cleaning

Data cleaning is the process of detecting and repairing corrupt or inaccurate records from a data set in order to improve the quality of data.

Data is generally considered high quality **if it is fit for its intended uses** in operations, decision making and planning. So, we will first determine what is the intended use of this dataset.

***Our goals:***

 1. Explore the the potential relationship between time (or location) and motor vehicle crashes.

 2. Draw Motor Vehicle Crash Distribution over different areas by zip codes, which could facilitate the government work.

Therefore, data we should pay attention to is about time and geographic location. We could ignore number of persons killed, number of persons injured, contributing factor vehicle and vehicle type.

### 1. Outliers

###### (1) Load zip code and borough mapping csv

- reference: https://www.nycbynatives.com/nyc_info/new_york_city_zip_codes.php

In [11]:
import pandas as pd

borough_name_mapping = {'Manhattan': 'MANHATTAN',
                       'Staten': 'STATEN ISLAND',
                       'Bronx': 'BRONX',
                       'Queens': 'QUEENS',
                       'Brooklyn': 'BROOKLYN'}

borough_zipcode_df = pd.read_csv('borough_zipcode.csv')

zipcodestr_to_borough_dict = dict()
zipcodes_to_borough_dict = dict()
for index, row in borough_zipcode_df.iterrows():
    zipcodestr_to_borough_dict[str(row['ZIP_CODE'])] = borough_name_mapping[row['BOROUGH']]
    zipcodes_to_borough_dict[row['ZIP_CODE']] = borough_name_mapping[row['BOROUGH']]

# To take a look at zipcode_to_borough_dict
zipcodes_to_borough_dict

{10001: 'MANHATTAN',
 10002: 'MANHATTAN',
 10003: 'MANHATTAN',
 10004: 'MANHATTAN',
 10005: 'MANHATTAN',
 10006: 'MANHATTAN',
 10007: 'MANHATTAN',
 10009: 'MANHATTAN',
 10010: 'MANHATTAN',
 10011: 'MANHATTAN',
 10012: 'MANHATTAN',
 10013: 'MANHATTAN',
 10014: 'MANHATTAN',
 10015: 'MANHATTAN',
 10016: 'MANHATTAN',
 10017: 'MANHATTAN',
 10018: 'MANHATTAN',
 10019: 'MANHATTAN',
 10020: 'MANHATTAN',
 10021: 'MANHATTAN',
 10022: 'MANHATTAN',
 10023: 'MANHATTAN',
 10024: 'MANHATTAN',
 10025: 'MANHATTAN',
 10026: 'MANHATTAN',
 10027: 'MANHATTAN',
 10028: 'MANHATTAN',
 10029: 'MANHATTAN',
 10030: 'MANHATTAN',
 10031: 'MANHATTAN',
 10032: 'MANHATTAN',
 10033: 'MANHATTAN',
 10034: 'MANHATTAN',
 10035: 'MANHATTAN',
 10036: 'MANHATTAN',
 10037: 'MANHATTAN',
 10038: 'MANHATTAN',
 10039: 'MANHATTAN',
 10040: 'MANHATTAN',
 10041: 'MANHATTAN',
 10044: 'MANHATTAN',
 10045: 'MANHATTAN',
 10048: 'MANHATTAN',
 10055: 'MANHATTAN',
 10060: 'MANHATTAN',
 10069: 'MANHATTAN',
 10090: 'MANHATTAN',
 10095: 'MANH

###### (2) BOROUGH

In [12]:
df_ol1 = ds.select('BOROUGH').update('BOROUGH', str.upper).to_df()

In [13]:
df_ol1['BOROUGH'].value_counts()

                 568098
BROOKLYN         400172
QUEENS           340459
MANHATTAN        292471
BRONX            184511
STATEN ISLAND     53545
Name: BOROUGH, dtype: int64

Therefore, except for null value, BOROUGH has no outliers.

###### (3) ZIP CODE

Given all the zip codes in NYC, check if all of them are in the range.

In [14]:
zipcodes_list = list(zipcodestr_to_borough_dict.keys())

In [15]:
df_ol2 = ds.select('ZIP CODE').to_df()

In [16]:
df_notin = df_ol2[~df_ol2['ZIP CODE'].isin(zipcodes_list)]
df_notin['ZIP CODE'].unique()

array(['', '10065', '11001', '10000', '11040', '10075', '11695', '11251',
       '11005', '10179', '     ', '10803'], dtype=object)

In [17]:
zipcode_change_dict = {}
for i in df_notin['ZIP CODE'].unique():
    zipcode_change_dict[i] = ''

In [18]:
# Find all the zip codes that are not in NYC
# They might be typo, so we need to check the zip code based on more specific geographic location information, 
# e.g. latitude,longtitude, on street name, cross street name, off street name
df_notin['ZIP CODE'].value_counts()

         568320
10065      8724
10075      3086
11001       547
11040       462
10000       348
11005       101
             42
11695        14
10803         2
10179         2
11251         2
Name: ZIP CODE, dtype: int64

***Update those rows whose ZIP CODE are outliers to empty string***

In [19]:
from openclean.function.eval.domain import Lookup
from openclean.function.eval.base import Col
ds_fix_zip = ds\
    .update('ZIP CODE', Lookup('ZIP CODE', zipcode_change_dict, default=Col('ZIP CODE')))
ds_fix_zip.head()

CRASH DATE CRASH TIME   BOROUGH ZIP CODE  LATITUDE  LONGITUDE  \
0  04/14/2021       5:32                                           
1  04/13/2021      21:35  BROOKLYN    11217  40.68358  -73.97617   
2  04/15/2021      16:15                                           
3  04/13/2021      16:00  BROOKLYN    11222                        
4  04/12/2021       8:25                            0          0   
5  04/13/2021      17:11                                           
6  04/13/2021      17:30    QUEENS    11106                        
7  04/16/2021      23:30                                           
8  04/11/2021      17:00                                           
9  04/16/2021      21:15                                           

                LOCATION                    ON STREET NAME CROSS STREET NAME  \
0                                  BRONX WHITESTONE BRIDGE                     
1  (40.68358, -73.97617)                                                       
2                                 HUTCHINSON RIVER PARKWAY                     
3                                        VANDERVORT AVENUE    ANTHONY STREET   
4             (0.0, 0.0)  EDSON AVENUE                                         
5                                   VERRAZANO BRIDGE UPPER                     
6                                                    33 st             31ave   
7                                            SHORE PARKWAY                     
8                         GOWANUS RAMP                                         
9                                 BRONX RIVER PARKWAY RAMP                     

                            OFF STREET NAME  ...  \
0                                            ...   
1  620       ATLANTIC AVENUE                 ...   
2                                            ...   
3                                            ...   
4                                            ...   
5                                            ...   
6                                            ...   
7                                            ...   
8                                            ...   
9                                            ...   

  CONTRIBUTING FACTOR VEHICLE 2 CONTRIBUTING FACTOR VEHICLE 3  \
0                   Unspecified                                 
1                                                               
2                                                               
3                   Unspecified                                 
4                   Unspecified                                 
5                   Unspecified                                 
6                   Unspecified                                 
7                                                               
8               Other Vehicular                                 
9                   Unspecified                                 

  CONTRIBUTING FACTOR VEHICLE 4 CONTRIBUTING FACTOR VEHICLE 5 COLLISION_ID  \
0                                                                  4407480   
1                                                                  4407147   
2                                                                  4407665   
3                                                                  4407811   
4                                                                  4406885   
5                                                                  4407883   
6                                                                  4408019   
7                                                                  4408060   
8                                                                  4406314   
9                                                                  4408149   

                   VEHICLE TYPE CODE 1 VEHICLE TYPE CODE 2  \
0                                Sedan               Sedan   
1                                Sedan                       
2  Station Wagon/Sport Utility Vehicle 

In [20]:
from openclean.operator.transform.filter import filter

df_test = ds_fix_zip\
    .filter(Col('ZIP CODE') == '11695')
df_test.head(20)

Empty DataFrame
Columns: [CRASH DATE, CRASH TIME, BOROUGH, ZIP CODE, LATITUDE, LONGITUDE, LOCATION, ON STREET NAME, CROSS STREET NAME, OFF STREET NAME, NUMBER OF PERSONS INJURED, NUMBER OF PERSONS KILLED, NUMBER OF PEDESTRIANS INJURED, NUMBER OF PEDESTRIANS KILLED, NUMBER OF CYCLIST INJURED, NUMBER OF CYCLIST KILLED, NUMBER OF MOTORIST INJURED, NUMBER OF MOTORIST KILLED, CONTRIBUTING FACTOR VEHICLE 1, CONTRIBUTING FACTOR VEHICLE 2, CONTRIBUTING FACTOR VEHICLE 3, CONTRIBUTING FACTOR VEHICLE 4, CONTRIBUTING FACTOR VEHICLE 5, COLLISION_ID, VEHICLE TYPE CODE 1, VEHICLE TYPE CODE 2, VEHICLE TYPE CODE 3, VEHICLE TYPE CODE 4, VEHICLE TYPE CODE 5]
Index: []

[0 rows x 29 columns]

Now, all the invalid zip code become empty string

### 2. Typos

* Find Typos in the columns of street names and standardize them

At first, we use the collision clustering algorithm to determine if there are any altenative representations of the street names. The specialied key-generating clustering algorithm USStreetNameKey is specific to identifying typos in street names. We apply this function on the columns 'ON STREET NAME', 'CROSS STREET NAME', 'OFF STREET NAME' to find clusters of alternative representations of a particular street name.



In [21]:
from openclean.cluster.key import KeyCollision
from openclean.function.value.key.fingerprint import Fingerprint
from openclean_geo.address.usstreet import USStreetNameKey

on_street_names = ds_fix_zip.update('ON STREET NAME', str.upper).distinct('ON STREET NAME')
cross_street_names = ds_fix_zip.update('CROSS STREET NAME', str.upper).distinct('CROSS STREET NAME')
off_street_names = ds_fix_zip.update('OFF STREET NAME', str.upper).distinct('OFF STREET NAME')

on_street_clusters = KeyCollision(func=USStreetNameKey()).clusters(on_street_names)
cross_street_clusters = KeyCollision(func=USStreetNameKey()).clusters(cross_street_names)
off_street_clusters = KeyCollision(func=USStreetNameKey()).clusters(off_street_names)

We then use clusters_print function to print the clusters found using the USStreetNameKey function.

In [22]:
def clusters_print(clusters, k=5):
    clusters = sorted(clusters, key=lambda x: len(x), reverse=True)
    val_count = sum([len(c) for c in clusters])
    print('Total number of clusters is {} with {} values'.format(len(clusters), val_count))
    for i in range(min(k, len(clusters))):
        print('\nCluster {}'.format(i + 1))
        for key, cnt in clusters[i].items():
            if key == '':
                key = "''"
            print(f'  {key} (x {cnt})')

In [23]:
print("Clusters in ON STREET NAMES")
clusters_print(on_street_clusters, 10)
print("\nClusters in CROSS STREET NAMES")
clusters_print(cross_street_clusters, 10)
print("\nClusters in OFF STREET NAMES")
clusters_print(off_street_clusters, 10)

Clusters in ON STREET NAMES
Total number of clusters is 3623 with 8431 values

Cluster 1
  FIRST AVENUE (x 1)
  1 AVENUE (x 125)
  1 AVENUE                         (x 5749)
  FIRST AVENUE                     (x 24)
  1ST AVENUE                       (x 21)
  FIRST AVE                        (x 3)
  1ST AVE                          (x 26)
  1 AVE                            (x 1)
   1ST AVENUE                      (x 1)

Cluster 2
  2 AVENUE (x 231)
  2 AVENUE                         (x 9029)
  2ND AVE (x 1)
  SECOND AVENUE                    (x 35)
  2 AVE                            (x 3)
  2ND AVE                          (x 5)
  SECOND AVE                       (x 1)
  2ND AVENUE                       (x 2)
  2 AV                             (x 1)

Cluster 3
  3 AVENUE (x 365)
  THIRD AVENUE (x 2)
  THIRD AVE (x 3)
  3 AVENUE                         (x 12486)
  THIRD AVENUE                     (x 260)
  THIRD AVE                        (x 7)
  3 AVE                            (x 6)
  

 We use the street specific standerdizer 'StandardizeUSStreetName' function to standerdize typos in the street names. Then we apply the generic FringerPrint clustering algorithm to determine if there are any remaining typos.

In [24]:
from openclean_geo.address.usstreet import StandardizeUSStreetName
f = StandardizeUSStreetName(characters='upper', alphanum=True, repeated=False)
on_street_names_std = f.apply(on_street_names, threads=3)
cross_street_names_std = f.apply(cross_street_names, threads=3)
off_street_names_std = f.apply(off_street_names, threads=3)

clusters_on_street = KeyCollision(func=Fingerprint(), threads=3).clusters(on_street_names_std)
clusters_cross_street = KeyCollision(func=Fingerprint(), threads=3).clusters(cross_street_names_std)
clusters_off_street = KeyCollision(func=Fingerprint(), threads=3).clusters(off_street_names_std)

print("Remaining Clusters in ON STREET NAMES")
clusters_print(clusters_on_street, 10)
print("\nRemaining Clusters in CROSS STREET NAMES")
clusters_print(clusters_cross_street, 10)
print("\nRemaining Clusters in OFF STREET NAMES")
clusters_print(clusters_off_street, 10)

Remaining Clusters in ON STREET NAMES
Total number of clusters is 20 with 41 values

Cluster 1
  ST NICHOLAS (x 5)
  ST NICHOLAS ST (x 1)
  NICHOLAS ST (x 27)

Cluster 2
  REV A JAMES POLITE AVE (x 1)
  REV JAMES A POLITE AVE (x 1)

Cluster 3
  EAST BEDFORD PARK BLVD (x 191)
  BEDFORD PARK BLVD EAST (x 4)

Cluster 4
  WEST BEDFORD PARK BLVD (x 51)
  BEDFORD PARK BLVD WEST (x 2)

Cluster 5
  NORTH BOUNDARY RD (x 1)
  BOUNDARY NORTH RD (x 1)

Cluster 6
  G C P L I E CDR (x 193)
  L I E G C P CDR (x 149)

Cluster 7
  GRAND CENTRAL PARKWAY SR WEST (x 1)
  GRAND CENTRAL PARKWAY WEST SR (x 1)

Cluster 8
  HUGH J GRANT CIR (x 37)
  HUGH GRANT J CIR (x 1)

Cluster 9
  CLEARVIEW EXPRESSWAY SR EAST (x 10)
  CLEARVIEW EXPRESSWAY EAST SR (x 10)

Cluster 10
  CROSS ISLAND PARKWAY SR SOUTH (x 32)
  CROSS ISLAND PARKWAY SOUTH SR (x 3)

Remaining Clusters in CROSS STREET NAMES
Total number of clusters is 19 with 38 values

Cluster 1
  2 1 AVE (x 1)
  2 AVE 1 AVE (x 1)

Cluster 2
  BAY 35 ST (x 49)
  3

We get rid of the remaining typos by using a map to store alternative versions of a street name as the keys and the actual street name as the value in the map. Then we iterate through the street name columns to replace the typos using the generated map.

In [25]:
ON_STREET_MAPPING= {}
OFF_STREET_MAPPING= {}
CROSS_STREET_MAPPING= {}

#Generates a map containing typos of ON STREET NAME column as keys
def create_on_street_dict(cluster):
    for i in range(len(cluster)):
        sorted_cluster = sorted(cluster[i].items(), key=lambda x: x[1])
        for j in range(len(sorted_cluster)):
            if j==0:
                value=sorted_cluster[j][0]
            else:
                ON_STREET_MAPPING[sorted_cluster[j][0]]=value

#Generates a map containing typos of CROSS STREET NAME column as keys
def create_cross_street_dict(cluster):
    for i in range(len(cluster)):
        sorted_cluster = sorted(cluster[i].items(), key=lambda x: x[1])
        for j in range(len(sorted_cluster)):
            if j==0:
                value=sorted_cluster[j][0]
            else:
                CROSS_STREET_MAPPING[sorted_cluster[j][0]]=value

#Generates a map containing typos of OFF STREET NAME column as keys
def create_off_street_dict(cluster):
    for i in range(len(cluster)):
        sorted_cluster = sorted(cluster[i].items(), key=lambda x: x[1])
        for j in range(len(sorted_cluster)):
            if j==0:
                value=sorted_cluster[j][0]
            else:
                OFF_STREET_MAPPING[sorted_cluster[j][0]]=value
                

In [26]:
#Functions to iterate through the columns and remove remaining typos using the generated maps

def remove_remaining_clusters_on_street(series): 
    on_street_name=series['ON STREET NAME']
    if on_street_name in ON_STREET_MAPPING.keys():
        return(ON_STREET_MAPPING[on_street_name])
    else:
        return on_street_name

def remove_remaining_clusters_cross_street(series):
    cross_street_name=series['CROSS STREET NAME']
    if cross_street_name in CROSS_STREET_MAPPING.keys():
        return(CROSS_STREET_MAPPING[cross_street_name])
    else:
        return cross_street_name

def remove_remaining_clusters_off_street(series):
    off_street_name=series['OFF STREET NAME']
    if off_street_name in OFF_STREET_MAPPING.keys():
        return(OFF_STREET_MAPPING[off_street_name])
    else:
        return off_street_name

In [27]:
# applying StandardizeUSStreetName function on columns.

ds_fix_street = ds.update('ON STREET NAME', f)
ds_fix_street = ds_fix_street.update('CROSS STREET NAME', f)
ds_fix_street = ds_fix_street.update('OFF STREET NAME', f)

columns = [
    'CRASH DATE',
    'CRASH TIME',
    'BOROUGH',
    'ZIP CODE',
    'LATITUDE',
    'LONGITUDE',
    'LOCATION',
    'ON STREET NAME',
    'CROSS STREET NAME',
    'OFF STREET NAME',
    'COLLISION_ID'
]

df_cluster_typos = ds_fix_street.select(columns).to_df()
df_cluster_typos.head(10)

CRASH DATE CRASH TIME   BOROUGH ZIP CODE  LATITUDE  LONGITUDE  \
0  04/14/2021       5:32                                           
1  04/13/2021      21:35  BROOKLYN    11217  40.68358  -73.97617   
2  04/15/2021      16:15                                           
3  04/13/2021      16:00  BROOKLYN    11222                        
4  04/12/2021       8:25                            0          0   
5  04/13/2021      17:11                                           
6  04/13/2021      17:30    QUEENS    11106                        
7  04/16/2021      23:30                                           
8  04/11/2021      17:00                                           
9  04/16/2021      21:15                                           

                LOCATION            ON STREET NAME CROSS STREET NAME  \
0                             BRONX WHITESTONE BRG                     
1  (40.68358, -73.97617)                                               
2                            HUTCHINSON RIVER PKWY                     
3                                   VANDERVORT AVE        ANTHONY ST   
4             (0.0, 0.0)                 EDSON AVE                     
5                           VERRAZANO BRIDGE UPPER                     
6                                            33 ST            31 AVE   
7                                       SHORE PKWY                     
8                                     GOWANUS RAMP                     
9                         BRONX RIVER PARKWAY RAMP                     

    OFF STREET NAME COLLISION_ID  
0                        4407480  
1  620 ATLANTIC AVE      4407147  
2                        4407665  
3                        4407811  
4                        4406885  
5                        4407883  
6                        4408019  
7                        4408060  
8                        4406314  
9                        4408149

In [28]:
#calling the functions to remove remaining typos

create_on_street_dict(clusters_on_street)
create_cross_street_dict(clusters_cross_street)
create_off_street_dict(clusters_off_street)

df_cluster_typos['ON STREET NAME']=df_cluster_typos.apply(remove_remaining_clusters_on_street,axis=1)
df_cluster_typos['CROSS STREET NAME']=df_cluster_typos.apply(remove_remaining_clusters_cross_street,axis=1)
df_cluster_typos['OFF STREET NAME']=df_cluster_typos.apply(remove_remaining_clusters_off_street,axis=1)

df_cluster_typos.head()
ds_fix_street_name=stream(df_cluster_typos)

### 3. Key violations

In this dataset, the primary key should be COLLISION_ID, because COLLISION_ID can uniquely identify each collision. We try to find if there is any key violation.

In [29]:
from openclean.operator.map.violations import key_violations

# Ensure that COLLISION_ID is a key candidate now.
groups = key_violations(df_cluster_typos, ['COLLISION_ID'])
assert len(groups) == 0

print('# of violations for KEY(COLLISION_ID) is {}'.format(len(groups)))

# Total number of businesses.
print('Total number of records is {}'.format(len(df_cluster_typos)))

# of violations for KEY(COLLISION_ID) is 0
Total number of records is 1839256


### 4. Missing values

The most important columns that we will use to analyze are "CRASH DATE, CRASH TIME, BOROUGH, ZIP CODE, LATITUDE, LONGITUDE", so we just need to repair missing values in these columns.

According to profiling results, we know that columns "CRASH DATE" and "CRASH TIME" do not have missing values so that we don't need to take them into consideration. Columns "LATITUDE" and "LONGITUDE" are consistent, which means that they either exist at the same time or miss at the same time, and thus we could consider them as a whole.

###### (1) Complete records

Find out records whose columns "CRASH DATE, CRASH TIME, BOROUGH, ZIP CODE, LATITUDE, LONGITUDE" are not empty.

In [30]:
from openclean.function.eval.logic import *
from openclean.function.eval.null import *
from openclean.function.eval.domain import *

COLUMNS = [
    'CRASH DATE',
    'CRASH TIME',
    'BOROUGH',
    'ZIP CODE',
    'LATITUDE',
    'LONGITUDE',
    'ON STREET NAME',
    'CROSS STREET NAME',
    'OFF STREET NAME',
    'COLLISION_ID'
]

data_complete = ds_fix_street_name\
    .select(columns=COLUMNS)\
    .where(And(IsNotEmpty('LATITUDE'),IsNotEmpty('LONGITUDE'),IsNotEmpty('BOROUGH'),IsNotEmpty('ZIP CODE')))\
    .to_df()

In [31]:
data_complete

CRASH DATE CRASH TIME        BOROUGH ZIP CODE    LATITUDE  \
1        04/13/2021      21:35       BROOKLYN    11217    40.68358   
13       05/21/2019      22:50       BROOKLYN    11201    40.69754   
15       02/26/2021      14:50          BRONX    10461   40.843464   
17       03/31/2021      22:20       BROOKLYN    11234   40.626457   
18       04/06/2021      22:58  STATEN ISLAND    10312   40.526894   
...             ...        ...            ...      ...         ...   
1839250  07/07/2012      17:35       BROOKLYN    11235  40.5758019   
1839251  07/06/2012      15:09      MANHATTAN    10035  40.8012354   
1839252  07/03/2012      17:30         QUEENS    11102  40.7747112   
1839253  07/01/2012      15:30       BROOKLYN    11236  40.6450318   
1839255  07/05/2012      23:00         QUEENS    11367  40.7199104   

           LONGITUDE ON STREET NAME CROSS STREET NAME     OFF STREET NAME  \
1          -73.97617                                     620 ATLANTIC AVE   
13         -73.98312        GOLD ST        CONCORD ST                       
15           -73.836                                   2819 MIDDLETOWN RD   
17           -73.918      RALPH AVE          AVENUE K                       
18         -74.16728    BARCLAY AVE        HYLAN BLVD                       
...              ...            ...               ...                 ...   
1839250  -73.9611371  BRIGHTON 6 ST    BRIGHTWATER CT                       
1839251  -73.9418153    EAST 119 ST          PARK AVE                       
1839252  -73.9333863         27 AVE              4 ST                       
1839253  -73.9199775      RALPH AVE      CLARENDON RD                       
1839255  -73.8090389   PARSONS BLVD        UNION TPKE                       

        COLLISION_ID  
1            4407147  
13           4136992  
15           4395664  
17           4403773  
18           4405244  
...              ...  
1839250       113584  
1839251        59654  
1839252       272592  
1839253       135041  
1839255       233172  

[1237781 rows x 10 columns]

There are over 1.2 million complete records which we don't need to repair missing values.

###### (2) Repair missing zipcode or borough using latitude and longitude

Find out records whose column "BOROUGH" or "ZIP CODE" is empty but "LATITUDE" and "LONGITUDE" are not empty.

In [32]:
data_2 = ds_fix_street_name\
    .select(columns=COLUMNS)\
    .where(And(IsNotEmpty('LATITUDE'),IsNotEmpty('LONGITUDE'),Or(IsEmpty('BOROUGH'),IsEmpty('ZIP CODE'))))\
    .to_df()

In [33]:
data_2

CRASH DATE CRASH TIME BOROUGH ZIP CODE    LATITUDE    LONGITUDE  \
4        04/12/2021       8:25                            0            0   
16       03/09/2021      11:00                    40.692547   -73.990974   
19       04/09/2021      14:45                    40.840775    -73.87246   
21       04/14/2021      11:00                    40.694035    -73.72679   
23       04/14/2021      14:40                    40.698807    -73.91837   
...             ...        ...     ...      ...         ...          ...   
1839202  07/03/2012      16:47                   40.8614057  -73.8277907   
1839212  07/04/2012      21:50                   40.6915381    -73.99911   
1839215  07/07/2012      18:40                   40.8673349  -73.8227066   
1839245  07/10/2012       7:00                   40.6276429  -73.8902159   
1839254  07/08/2012      18:30                   40.7861217  -73.8040782   

            ON STREET NAME CROSS STREET NAME OFF STREET NAME COLLISION_ID  
4                EDSON AVE                                        4406885  
16                COURT ST      JORALEMON ST                      4397513  
19        BRONX RIVER PKWY                                        4405914  
21       CROSS ISLAND PKWY                                        4407366  
23              MYRTLE AVE                                        4407461  
...                    ...               ...             ...          ...  
1839202                                                           2912100  
1839212                                                           2983048  
1839215                                                           2912116  
1839245                                                           2960052  
1839254                                                           3055617  

[386489 rows x 10 columns]

There are 0.38 million records.

**In order to repair missing values in this part, we could use geographic coordinates(latitude and longitude) to obtain a zipcode, and then transform a zipcode to a borough.**

First, we generate a csv file containing LATITUDE,LONGITUDE and COLLISION_ID.

In [34]:
data_togeo = data_2.drop(['CRASH DATE','CRASH TIME','BOROUGH','ZIP CODE','ON STREET NAME','CROSS STREET NAME','OFF STREET NAME'],axis=1)
data_togeo.to_csv('geo_coordinates.csv')

We obtain a new csv file - "out_test.csv"

In [35]:
import pandas as pd
geo = pd.read_csv('out_test.csv')
geo = geo[['ZIP_CODE']]

In [36]:
data_2 = data_2.reset_index()
data_2 = data_2.drop(['index'],axis = 1)
data_2 = pd.concat([data_2,geo],axis=1)
data_2.loc[(data_2['ZIP CODE'] == ''),'ZIP CODE'] = data_2['ZIP_CODE']
data_2.loc[(data_2['ZIP CODE'] == 'None'),'ZIP CODE'] = ''
data_2=data_2.drop(['ZIP_CODE'],axis=1)
data_2

CRASH DATE CRASH TIME BOROUGH ZIP CODE    LATITUDE    LONGITUDE  \
0       04/12/2021       8:25                            0            0   
1       03/09/2021      11:00            11201   40.692547   -73.990974   
2       04/09/2021      14:45            10460   40.840775    -73.87246   
3       04/14/2021      11:00            11411   40.694035    -73.72679   
4       04/14/2021      14:40            11237   40.698807    -73.91837   
...            ...        ...     ...      ...         ...          ...   
386484  07/03/2012      16:47            10475  40.8614057  -73.8277907   
386485  07/04/2012      21:50            11201  40.6915381    -73.99911   
386486  07/07/2012      18:40            10475  40.8673349  -73.8227066   
386487  07/10/2012       7:00            11236  40.6276429  -73.8902159   
386488  07/08/2012      18:30            11357  40.7861217  -73.8040782   

           ON STREET NAME CROSS STREET NAME OFF STREET NAME COLLISION_ID  
0               EDSON AVE                                        4406885  
1                COURT ST      JORALEMON ST                      4397513  
2        BRONX RIVER PKWY                                        4405914  
3       CROSS ISLAND PKWY                                        4407366  
4              MYRTLE AVE                                        4407461  
...                   ...               ...             ...          ...  
386484                                                           2912100  
386485                                                           2983048  
386486                                                           2912116  
386487                                                           2960052  
386488                                                           3055617  

[386489 rows x 10 columns]

In [37]:
manhattan_zipcode = ['10001','10002','10003','10004','10005','10006','10007','10008','10009','10010','10011','10012','10013','10014','10015','10016','10017','10018','10019','10020','10021','10022','10023','10024','10025','10026','10027','10028','10029','10030','10031','10032','10033','10034','10035','10036','10037','10038','10039','10040','10041','10043','10044','10045','10048','10055','10060','10065','10069','10075','10080','10081','10087','10090','10095','10098','10099','10103','10104','10105','10106','10107','10110','10111','10112','10115','10118','10119','10120','10121','10122','10123','10128','10151','10152','10153','10154','10155','10158','10161','10162','10165','10166','10167','10168','10169','10170','10171','10172','10173','10174','10175','10176','10177','10178','10199','10270','10271','10278','10279','10280','10281','10282']
staten_zipcode = ['10301','10302','10303','10304','10305','10306','10307','10308','10309','10310','10311','10312','10313','10314']
bronx_zipcode = ['10451','10452','10453','10454','10455','10456','10457','10458','10459','10460','10461','10462','10463','10464','10465','10466','10467','10468','10469','10470','10471','10472','10473','10474','10475']
queens_zipcode = ['11004','11005','11101','11102','11103','11104','11105','11106','11109','11120','11351','11352','11354','11355','11356','11357','11358','11359','11360','11361','11362','11363','11364','11365','11366','11367','11368','11369','11370','11371','11372','11373','11374','11375','11377','11378','11379','11380','11381','11385','11386','11405','11411','11412','11413','11414','11415','11416','11417','11418','11419','11420','11421','11422','11423','11424','11425','11426','11427','11428','11429','11430','11431','11432','11433','11434','11435','11436','11437','11439','11451','11499','11690','11691','11692','11693','11694','11695','11697']
brooklyn_zipcode = ['11201','11202','11203','11204','11205','11206','11207','11208','11209','11210','11211','11212','11213','11214','11215','11216','11217','11218','11219','11220','11221','11222','11223','11224','11225','11226','11228','11229','11230','11231','11232','11233','11234','11235','11236','11237','11238','11239','11241','11242','11243','11245','11247','11249','11251','11252','11256']

In [38]:
def zipcode2borough(series):
        borough = series['BOROUGH']
        zipcode = series['ZIP CODE']
      
        if borough == '':
            if zipcode in manhattan_zipcode:
                borough = 'MANHATTAN'
            elif zipcode in staten_zipcode:
                borough = 'STATEN ISLAND'
            elif zipcode in bronx_zipcode:
                borough = 'BRONX'
            elif zipcode in queens_zipcode:
                borough = 'QUEENS'
            elif zipcode in brooklyn_zipcode:
                borough = 'BROOKLYN'
            else:
                borough = ''
        return borough

In [39]:
data_2['BOROUGH'] = data_2.apply(zipcode2borough,axis=1)

In [40]:
data_2

CRASH DATE CRASH TIME   BOROUGH ZIP CODE    LATITUDE    LONGITUDE  \
0       04/12/2021       8:25                              0            0   
1       03/09/2021      11:00  BROOKLYN    11201   40.692547   -73.990974   
2       04/09/2021      14:45     BRONX    10460   40.840775    -73.87246   
3       04/14/2021      11:00    QUEENS    11411   40.694035    -73.72679   
4       04/14/2021      14:40  BROOKLYN    11237   40.698807    -73.91837   
...            ...        ...       ...      ...         ...          ...   
386484  07/03/2012      16:47     BRONX    10475  40.8614057  -73.8277907   
386485  07/04/2012      21:50  BROOKLYN    11201  40.6915381    -73.99911   
386486  07/07/2012      18:40     BRONX    10475  40.8673349  -73.8227066   
386487  07/10/2012       7:00  BROOKLYN    11236  40.6276429  -73.8902159   
386488  07/08/2012      18:30    QUEENS    11357  40.7861217  -73.8040782   

           ON STREET NAME CROSS STREET NAME OFF STREET NAME COLLISION_ID  
0               EDSON AVE                                        4406885  
1                COURT ST      JORALEMON ST                      4397513  
2        BRONX RIVER PKWY                                        4405914  
3       CROSS ISLAND PKWY                                        4407366  
4              MYRTLE AVE                                        4407461  
...                   ...               ...             ...          ...  
386484                                                           2912100  
386485                                                           2983048  
386486                                                           2912116  
386487                                                           2960052  
386488                                                           3055617  

[386489 rows x 10 columns]

###### （3） Missing values that we could not repair

In [41]:
data_3 = ds_fix_street_name\
    .select(columns=COLUMNS)\
    .where(And(IsEmpty('LATITUDE'),IsEmpty('LONGITUDE')))\
    .to_df()

In [42]:
data_3

CRASH DATE CRASH TIME   BOROUGH ZIP CODE LATITUDE LONGITUDE  \
0        04/14/2021       5:32                                         
2        04/15/2021      16:15                                         
3        04/13/2021      16:00  BROOKLYN    11222                      
5        04/13/2021      17:11                                         
6        04/13/2021      17:30    QUEENS    11106                      
...             ...        ...       ...      ...      ...       ...   
1839189  07/09/2012       5:15                                         
1839208  07/02/2012      15:20                                         
1839223  07/09/2012      21:19                                         
1839238  07/06/2012      16:10                                         
1839249  07/08/2012       1:30                                         

                 ON STREET NAME CROSS STREET NAME  \
0          BRONX WHITESTONE BRG                     
2         HUTCHINSON RIVER PKWY                     
3                VANDERVORT AVE        ANTHONY ST   
5        VERRAZANO BRIDGE UPPER                     
6                         33 ST            31 AVE   
...                         ...               ...   
1839189                                             
1839208              OCEAN PKWY        SHORE PKWY   
1839223                                             
1839238                                             
1839249            DIVISION AVE         MARCY AVE   

                        OFF STREET NAME COLLISION_ID  
0                                            4407480  
2                                            4407665  
3                                            4407811  
5                                            4407883  
6                                            4408019  
...                                 ...          ...  
1839189  2601 OCEAN PARKWAY PARKING LOT       113583  
1839208                                       113554  
1839223                                      2937609  
1839238                                      2888480  
1839249                                       193887  

[214986 rows x 10 columns]

In [43]:
dataframes = [data_complete,data_2,data_3]
dataset_fix_missing = pd.concat(dataframes)

### 5. Inconsistency between values

###### (1)  Find violations of the functional dependency: ZIP CODE => BOROUGH.

In [44]:
from openclean.function.eval.logic import Or
from openclean.function.eval.null import IsEmpty
from openclean.operator.transform.filter import delete

df_fd2 = delete(dataset_fix_missing, Or(IsEmpty('ZIP CODE'), IsEmpty('BOROUGH')))

In [45]:
# Find violations of the functional dependency: ZIP CODE -> BOROUGH.

from openclean.operator.map.violations import fd_violations
fd2_violations = fd_violations(df_fd2, lhs='ZIP CODE', rhs='BOROUGH')
print('# of violations for FD(ZIP CODE -> BOROUGH) is {}'.format(len(fd2_violations)))

# of violations for FD(ZIP CODE -> BOROUGH) is 6


In [46]:
# Show street names that cause violations of the functional dependency.

from openclean.operator.collector.count import distinct

print('ZIP CODE | BOROUGH')
print('=============|===============')
for key in fd2_violations:
    conflicts = distinct(fd2_violations.get(key), 'BOROUGH').most_common()
    street, count = conflicts[0]
    print('{:<12} | {} x {}'.format(key, count, street))
    for street, count in conflicts[1:]:
        print('             | {} x {}'.format(count, street))
    print('-------------|---------------')

ZIP CODE | BOROUGH
=============|===============
11237        | 8547 x BROOKLYN
             | 325 x QUEENS
-------------|---------------
11385        | 18473 x QUEENS
             | 34 x BROOKLYN
-------------|---------------
11222        | 12351 x BROOKLYN
             | 1 x MANHATTAN
-------------|---------------
10301        | 7664 x STATEN ISLAND
             | 2 x QUEENS
-------------|---------------
11208        | 18743 x BROOKLYN
             | 171 x QUEENS
-------------|---------------
11421        | 6189 x QUEENS
             | 7 x BROOKLYN
-------------|---------------


Find the correct BOROUGH and repair them.

In [47]:
print(zipcodestr_to_borough_dict['11237'])
print(zipcodestr_to_borough_dict['11385'])
print(zipcodestr_to_borough_dict['11222'])
print(zipcodestr_to_borough_dict['10301'])
print(zipcodestr_to_borough_dict['11208'])
print(zipcodestr_to_borough_dict['11421'])

BROOKLYN
QUEENS
BROOKLYN
STATEN ISLAND
BROOKLYN
QUEENS


In [48]:
dataset_fix_missing.loc[(dataset_fix_missing['ZIP CODE']=='11237')&(dataset_fix_missing['BOROUGH']=='QUEENS'),'BOROUGH'] = 'BROOKLYN'

In [49]:
dataset_fix_missing.loc[(dataset_fix_missing['ZIP CODE']=='11385')&(dataset_fix_missing['BOROUGH']=='BROOKLYN'),'BOROUGH'] = 'QUEENS'

In [50]:
dataset_fix_missing.loc[(dataset_fix_missing['ZIP CODE']=='11222')&(dataset_fix_missing['BOROUGH']=='MANHATTAN'),'BOROUGH'] = 'BROOKLYN'

In [51]:
dataset_fix_missing.loc[(dataset_fix_missing['ZIP CODE']=='10301')&(dataset_fix_missing['BOROUGH']=='QUEENS'),'BOROUGH'] = 'BROOKLYN'

In [52]:
dataset_fix_missing.loc[(dataset_fix_missing['ZIP CODE']=='11421')&(dataset_fix_missing['BOROUGH']=='BROOKLYN'),'BOROUGH'] = 'QUEENS'

###### (2) Find violations of the functional dependency: [OFF STREET NAME, BOROUGH] -> ZIP CODE.

In [53]:
df_fd1 = delete(dataset_fix_missing, Or(IsEmpty('OFF STREET NAME'), IsEmpty('BOROUGH'), IsEmpty('ZIP CODE')))

# What is the pandas equivalent ??? These are the rows we want to delete:
# df[df['OFF STREET NAME'].isnull() | df['BOROUGH'].isnull() | df['ZIP CODE'].isnull()]

In [54]:
# Find violations of the functional dependency: [OFF STREET NAME, BOROUGH] -> ZIP CODE.

fd1_violations = fd_violations(df_fd1, lhs=['OFF STREET NAME', 'BOROUGH'], rhs='ZIP CODE')
print('# of violations for FD(OFF STREET NAME, BOROUGH -> ZIP CODE) is {}'.format(len(fd1_violations)))

# of violations for FD(OFF STREET NAME, BOROUGH -> ZIP CODE) is 901


In [55]:
# Show street names that cause violations of the functional dependency.

from openclean.operator.collector.count import distinct

print('OFF STREET NAME，  BOROUGH | ZIP CODE')
print('============================|===============')
for key in fd1_violations:
    conflicts = distinct(fd1_violations.get(key), ['ZIP CODE']).most_common()
    street, count = conflicts[0]
    print('{} | {} x {}'.format(key, count, street))
    for street, count in conflicts[1:]:
        print('                             | {} x {}'.format(count, street))
    print('-------------|---------------')

OFF STREET NAME，  BOROUGH | ZIP CODE
============================|===============
('60 74 70 ST', 'QUEENS') | 1 x 11378
                             | 1 x 11379
-------------|---------------
('455 GATEWAY DR', 'BROOKLYN') | 9 x 11239
                             | 1 x 11208
-------------|---------------
('104 45 32 AVE', 'QUEENS') | 3 x 11368
                             | 1 x 11369
-------------|---------------
('37 19 37 ST', 'QUEENS') | 5 x 11105
                             | 1 x 11101
-------------|---------------
('41 22 50 ST', 'QUEENS') | 1 x 11377
                             | 1 x 11105
-------------|---------------
('39 WEST 32 ST', 'MANHATTAN') | 3 x 10018
                             | 2 x 10001
-------------|---------------
('21 06 35 ST', 'QUEENS') | 4 x 11106
                             | 1 x 11105
-------------|---------------
('WEST 51 ST', 'MANHATTAN') | 13 x 10019
                             | 1 x 10111
-------------|---------------
('25 95 37 ST', 'QUEENS') | 2 x

('23 WEST 43 ST', 'MANHATTAN') | 2 x 10010
                             | 1 x 10036
-------------|---------------
('149 EAST 149 ST', 'BRONX') | 2 x 10451
                             | 1 x 10455
-------------|---------------
('30 EAST 10 ST', 'MANHATTAN') | 3 x 10016
                             | 1 x 10003
-------------|---------------
('112 EAST 65 ST', 'MANHATTAN') | 6 x 10029
                             | 1 x 10065
-------------|---------------
('21 23 31 AVE', 'QUEENS') | 1 x 11106
                             | 1 x 11105
-------------|---------------
('30 55 84 ST', 'QUEENS') | 1 x 11370
                             | 1 x 11373
-------------|---------------
('30 88 44 ST', 'QUEENS') | 1 x 11103
                             | 1 x 11369
-------------|---------------
('125 EAST 116 ST', 'MANHATTAN') | 19 x 10035
                             | 1 x 10029
-------------|---------------
('1065 ELTON ST', 'BROOKLYN') | 4 x 11208
                             | 1 x 11239
-------------|---

('65 WEST 36 ST', 'MANHATTAN') | 1 x 10018
                             | 1 x 10023
-------------|---------------
('132 WEST 45 ST', 'MANHATTAN') | 8 x 10037
                             | 1 x 10036
-------------|---------------
('40 WEST 8 ST', 'MANHATTAN') | 3 x 10018
                             | 1 x 10011
-------------|---------------
('31 WEST 125 ST', 'MANHATTAN') | 5 x 10001
                             | 2 x 10027
-------------|---------------
('34 24 98 ST', 'QUEENS') | 2 x 11369
                             | 1 x 11368
-------------|---------------
('21 77 33 ST', 'QUEENS') | 1 x 11105
                             | 1 x 11370
-------------|---------------
('27 WEST 105 ST', 'MANHATTAN') | 2 x 10001
                             | 1 x 10025
-------------|---------------
('30 37 81 ST', 'QUEENS') | 2 x 11372
                             | 1 x 11370
-------------|---------------
('WEST 75 ST', 'MANHATTAN') | 7 x 10006
                             | 3 x 10023
-------------|------

('53 WEST 53 ST', 'MANHATTAN') | 5 x 10019
                             | 1 x 10103
-------------|---------------
('18 WEST 23 ST', 'MANHATTAN') | 3 x 10010
                             | 3 x 10011
-------------|---------------
('30 EAST 44 ST', 'MANHATTAN') | 2 x 10016
                             | 1 x 10017
-------------|---------------
('123 WEST 117 ST', 'MANHATTAN') | 5 x 10027
                             | 1 x 10026
-------------|---------------
('16 EAST 30 ST', 'MANHATTAN') | 2 x 10003
                             | 1 x 10016
-------------|---------------
('31 WEST 19 ST', 'MANHATTAN') | 4 x 10001
                             | 1 x 10011
-------------|---------------
('35 EAST 1 ST', 'MANHATTAN') | 6 x 10016
                             | 2 x 10003
-------------|---------------
('25 37 97 ST', 'QUEENS') | 1 x 11369
                             | 1 x 11368
-------------|---------------
('26 WEST 38 ST', 'MANHATTAN') | 4 x 10010
                             | 1 x 10018
--------

('50 24 98 ST', 'QUEENS') | 1 x 11368
                             | 1 x 11369
-------------|---------------
('39 WEST 31 ST', 'MANHATTAN') | 1 x 10001
                             | 1 x 10018
-------------|---------------
('25 37 100 ST', 'QUEENS') | 1 x 11369
                             | 1 x 11368
-------------|---------------
('60 41 72 ST', 'QUEENS') | 2 x 11377
                             | 1 x 11378
-------------|---------------
('22 94 ST', 'QUEENS') | 2 x 11416
                             | 1 x 11369
-------------|---------------
('21 EAST 47 ST', 'MANHATTAN') | 1 x 10017
                             | 1 x 10010
-------------|---------------
('49 WEST 49 ST', 'MANHATTAN') | 5 x 10019
                             | 2 x 10036
                             | 1 x 10112
-------------|---------------
('59 WEST ST', 'MANHATTAN') | 2 x 10019
                             | 1 x 10006
-------------|---------------
('WEST 139 ST', 'MANHATTAN') | 3 x 10007
                             | 

('125 EAST 51 ST', 'MANHATTAN') | 1 x 10022
                             | 1 x 10035
-------------|---------------
('115 WEST 141 ST', 'MANHATTAN') | 2 x 10026
                             | 1 x 10030
-------------|---------------
('128 WEST 10 ST', 'MANHATTAN') | 1 x 10014
                             | 1 x 10027
-------------|---------------
('2 EAST 24 ST', 'MANHATTAN') | 1 x 10010
                             | 1 x 10003
-------------|---------------
('30 EAST 127 ST', 'MANHATTAN') | 1 x 10035
                             | 1 x 10016
-------------|---------------
('12 WEST 40 ST', 'MANHATTAN') | 1 x 10018
                             | 1 x 10011
-------------|---------------
('19 WEST 110 ST', 'MANHATTAN') | 1 x 10026
                             | 1 x 10011
-------------|---------------
('16 WEST 46 ST', 'MANHATTAN') | 1 x 10036
                             | 1 x 10011
-------------|---------------
('41 EAST 41 ST', 'MANHATTAN') | 5 x 10017
                             | 1 x 10016

('2 AVE', 'MANHATTAN') | 25 x 10016
                             | 18 x 10022
                             | 17 x 10029
                             | 13 x 10017
                             | 9 x 10003
                             | 5 x 10065
                             | 3 x 10021
                             | 3 x 10075
                             | 3 x 10010
                             | 2 x 10128
                             | 2 x 10028
-------------|---------------
('PARK PLACE', 'BROOKLYN') | 1 x 11238
                             | 1 x 11216
-------------|---------------
('PACIFIC ST', 'BROOKLYN') | 2 x 11216
                             | 2 x 11238
                             | 2 x 11233
-------------|---------------
('AVENUE P', 'BROOKLYN') | 5 x 11230
                             | 5 x 11223
                             | 2 x 11229
                             | 1 x 11214
-------------|---------------
('WOODHAVEN BLVD', 'QUEENS') | 5 x 11416
                             

('10 AVE', 'MANHATTAN') | 9 x 10036
                             | 7 x 10019
                             | 5 x 10001
                             | 1 x 10018
                             | 1 x 10011
-------------|---------------
('NORTHERN BLVD', 'QUEENS') | 23 x 11361
                             | 15 x 11358
                             | 12 x 11363
                             | 11 x 11101
                             | 10 x 11362
                             | 5 x 11354
                             | 3 x 11377
                             | 2 x 11103
                             | 2 x 11369
                             | 2 x 11372
-------------|---------------
('190 ST', 'QUEENS') | 2 x 11412
                             | 1 x 11358
-------------|---------------
('AVENUE L', 'BROOKLYN') | 7 x 11230
                             | 2 x 11236
-------------|---------------
('BEDFORD AVE', 'BROOKLYN') | 4 x 11216
                             | 3 x 11211
                             | 3 

('WEST 141 ST', 'MANHATTAN') | 2 x 10030
                             | 1 x 10031
                             | 1 x 10027
-------------|---------------
('48 ST', 'BROOKLYN') | 6 x 11220
                             | 5 x 11219
                             | 1 x 11232
-------------|---------------
('58 ST', 'BROOKLYN') | 6 x 11219
                             | 5 x 11220
                             | 2 x 11204
-------------|---------------
('AVENUE O', 'BROOKLYN') | 10 x 11230
                             | 1 x 11204
-------------|---------------
('92 ST', 'QUEENS') | 2 x 11373
                             | 1 x 11372
-------------|---------------
('EAST 2 ST', 'BROOKLYN') | 4 x 11230
                             | 4 x 11218
-------------|---------------
('60 ST', 'BROOKLYN') | 13 x 11204
                             | 8 x 11220
                             | 8 x 11219
-------------|---------------
('42 ST', 'BROOKLYN') | 7 x 11219
                             | 1 x 11204
------------

('11 AVE', 'BROOKLYN') | 6 x 11219
                             | 1 x 11228
-------------|---------------
('MAIN ST', 'QUEENS') | 4 x 11367
                             | 3 x 11355
                             | 2 x 11435
                             | 1 x 11354
-------------|---------------
('31 ST', 'QUEENS') | 3 x 11102
                             | 2 x 11101
                             | 2 x 11106
                             | 1 x 11105
-------------|---------------
('AVENUE K', 'BROOKLYN') | 2 x 11230
                             | 1 x 11236
-------------|---------------
('DEAN ST', 'BROOKLYN') | 2 x 11217
                             | 1 x 11233
                             | 1 x 11238
-------------|---------------
('40 ST', 'BROOKLYN') | 2 x 11218
                             | 1 x 11219
-------------|---------------
('45 ST', 'BROOKLYN') | 3 x 11219
                             | 1 x 11220
-------------|---------------
('19 AVE', 'BROOKLYN') | 5 x 11204
                     

('FDR DR', 'MANHATTAN') | 7 x 10010
                             | 2 x 10002
                             | 2 x 10017
                             | 2 x 10016
                             | 1 x 10128
-------------|---------------
('AVENUE D', 'BROOKLYN') | 1 x 11226
                             | 1 x 11236
-------------|---------------
('214 PLACE', 'QUEENS') | 2 x 11361
                             | 1 x 11360
-------------|---------------
('BROADWAY', 'BRONX') | 5 x 10463
                             | 1 x 10471
-------------|---------------
('40 AVE', 'QUEENS') | 3 x 11101
                             | 1 x 11361
-------------|---------------
('1 WATERSIDE PLZ', 'BROOKLYN') | 1 x 11235
                             | 1 x 11205
-------------|---------------
('46 AVE', 'QUEENS') | 2 x 11358
                             | 1 x 11361
-------------|---------------
('EAST HOUSTON ST', 'MANHATTAN') | 8 x 10002
                             | 3 x 10012
                             | 1 x 10009


('18 ST', 'BROOKLYN') | 1 x 11215
                             | 1 x 11232
-------------|---------------
('HAMILTON AVE', 'BROOKLYN') | 4 x 11215
                             | 1 x 11231
-------------|---------------
('FORT WASHINGTON AVE', 'MANHATTAN') | 2 x 10032
                             | 1 x 10040
-------------|---------------
('12 AVE', 'MANHATTAN') | 3 x 10019
                             | 1 x 10027
-------------|---------------
('32 ST', 'QUEENS') | 1 x 11106
                             | 1 x 11102
-------------|---------------
('205 ST', 'QUEENS') | 1 x 11361
                             | 1 x 11412
-------------|---------------
('EAST 49 ST', 'MANHATTAN') | 5 x 10017
                             | 4 x 10171
-------------|---------------
('SHERIDAN AVE', 'BRONX') | 1 x 10456
                             | 1 x 10451
-------------|---------------
('DYCKMAN ST', 'MANHATTAN') | 2 x 10033
                             | 2 x 10040
-------------|---------------
('195 ST', 'QUEENS

('COURT ST', 'BROOKLYN') | 2 x 11201
                             | 1 x 11231
-------------|---------------
('SOUTHERN BLVD', 'BRONX') | 3 x 10460
                             | 2 x 10458
-------------|---------------
('EAST 180 ST', 'BRONX') | 1 x 10460
                             | 1 x 10457
-------------|---------------
('EAST 14 ST', 'MANHATTAN') | 8 x 10003
                             | 1 x 10009
-------------|---------------
('NEPTUNE AVE', 'BROOKLYN') | 3 x 11224
                             | 2 x 11235
-------------|---------------
('60 AVE', 'QUEENS') | 1 x 11378
                             | 1 x 11373
-------------|---------------
('41 RD', 'QUEENS') | 1 x 11355
                             | 1 x 11101
-------------|---------------
('COLLEGE POINT BLVD', 'QUEENS') | 5 x 11354
                             | 5 x 11355
                             | 1 x 11368
-------------|---------------
('48 ST', 'QUEENS') | 3 x 11378
                             | 1 x 11103
-------------|-

('EAST 39 ST', 'MANHATTAN') | 3 x 10016
                             | 1 x 10018
-------------|---------------
('EMPIRE BLVD', 'BROOKLYN') | 5 x 11225
                             | 2 x 11213
-------------|---------------
('118 ST', 'QUEENS') | 1 x 11415
                             | 1 x 11419
-------------|---------------
('71 AVE', 'QUEENS') | 1 x 11375
                             | 1 x 11385
-------------|---------------
('222 ST', 'QUEENS') | 1 x 11429
                             | 1 x 11428
-------------|---------------
('BLAKE AVE', 'BROOKLYN') | 4 x 11212
                             | 1 x 11208
-------------|---------------
('69 ST', 'QUEENS') | 1 x 11379
                             | 1 x 11377
-------------|---------------
('VALENTINE AVE', 'BRONX') | 2 x 10458
                             | 1 x 10457
-------------|---------------
('WOODSIDE AVE', 'QUEENS') | 2 x 11377
                             | 1 x 11373
                             | 1 x 11104
-------------|---------

In [58]:
#for key in fd1_violations:
#    conflicts = distinct(fd1_violations.get(key), ['ZIP CODE']).most_common()
#    dataset_fix_missing.loc[(dataset_fix_missing['OFF STREET NAME']==key[0])&(dataset_fix_missing['BOROUGH']==key[1]),'ZIP CODE'] = ''
cleaned_data = dataset_fix_missing.reset_index()

In [59]:
cleaned_data

index  CRASH DATE CRASH TIME        BOROUGH ZIP CODE   LATITUDE  \
0              1  04/13/2021      21:35       BROOKLYN    11217   40.68358   
1             13  05/21/2019      22:50       BROOKLYN    11201   40.69754   
2             15  02/26/2021      14:50          BRONX    10461  40.843464   
3             17  03/31/2021      22:20       BROOKLYN    11234  40.626457   
4             18  04/06/2021      22:58  STATEN ISLAND    10312  40.526894   
...          ...         ...        ...            ...      ...        ...   
1839251  1839189  07/09/2012       5:15                                      
1839252  1839208  07/02/2012      15:20                                      
1839253  1839223  07/09/2012      21:19                                      
1839254  1839238  07/06/2012      16:10                                      
1839255  1839249  07/08/2012       1:30                                      

         LONGITUDE ON STREET NAME CROSS STREET NAME  \
0        -73.97617                                    
1        -73.98312        GOLD ST        CONCORD ST   
2          -73.836                                    
3          -73.918      RALPH AVE          AVENUE K   
4        -74.16728    BARCLAY AVE        HYLAN BLVD   
...            ...            ...               ...   
1839251                                               
1839252                OCEAN PKWY        SHORE PKWY   
1839253                                               
1839254                                               
1839255              DIVISION AVE         MARCY AVE   

                        OFF STREET NAME COLLISION_ID  
0                      620 ATLANTIC AVE      4407147  
1                                            4136992  
2                    2819 MIDDLETOWN RD      4395664  
3                                            4403773  
4                                            4405244  
...                                 ...          ...  
1839251  2601 OCEAN PARKWAY PARKING LOT       113583  
1839252                                       113554  
1839253                                      2937609  
1839254                                      2888480  
1839255                                       193887  

[1839256 rows x 11 columns]

In [60]:
cleaned_data.to_csv('cleaned_data.csv')

In [ ]:
cleaned_data